chain/jsonrpc/client/src/lib.rs

```diff
@@ -41,32 +42,42 @@ where
     R: serde::de::DeserializeOwned + 'static,
 {
     let request = Message::request(method.to_string(), serde_json::to_value(&params).unwrap());
-    // TODO: simplify this.
-    client
-        .post(server_addr)
-        .insert_header(("Content-Type", "application/json"))
-        .send_json(&request)
-        .map_err(|err| RpcError::new_internal_error(None, format!("{:?}", err)))
-        .and_then(|mut response| {
-            response.body().limit(PAYLOAD_LIMIT).map(|body| match body {
-                Ok(bytes) => from_slice(&bytes).map_err(|err| {
-                    RpcError::parse_error(format!("Error {:?} in {:?}", err, bytes))
-                }),
-                Err(err) => {
-                    Err(RpcError::parse_error(format!("Failed to retrieve payload: {:?}", err)))
-                }
-            })
-        })
-        .and_then(|message| {
-            future::ready(match message {
-                Message::Response(resp) => resp.result.and_then(|x| {
-                    serde_json::from_value(x)
-                        .map_err(|err| RpcError::parse_error(format!("Failed to parse: {:?}", err)))
-                }),
-                _ => Err(RpcError::parse_error("Failed to parse JSON RPC response".to_string())),
-            })
-        })
-        .boxed_local()
+    let client = client.clone();
+    let server_addr = server_addr.to_string();
+
+    async move {
+        let response = client
+            .post(&server_addr)
+            .header("Content-Type", "application/json")
+            .json(&request)
+            .send()
+            .await
+            .map_err(|err| RpcError::new_internal_error(None, format!("{:?}", err)))?;
+
+        let bytes = response.bytes().await.map_err(|err| {
+            RpcError::parse_error(format!("Failed to retrieve payload: {:?}", err))
+        })?;
+
+        if bytes.len() > PAYLOAD_LIMIT {
+            return Err(RpcError::parse_error(format!(
+                "Response payload too large: {} bytes, limit: {} bytes",
+                bytes.len(),
+                PAYLOAD_LIMIT
+            )));
+        }
+
+        let message = from_slice(&bytes)
+            .map_err(|err| RpcError::parse_error(format!("Error {:?} in {:?}", err, bytes)))?;
+
+        match message {
+            Message::Response(resp) => resp.result.and_then(|x| {
+                serde_json::from_value(x)
+                    .map_err(|err| RpcError::parse_error(format!("Failed to parse: {:?}", err)))
+            }),
+            _ => Err(RpcError::parse_error("Failed to parse JSON RPC response".to_string())),
+        }
+    }
+    .boxed_local()
```

Comment:
Could we get rid of all the boxed_local and use boxed() and BoxFuture? The local ones are not send and we'll eventually need to get rid of them.

Reply:
Sounds good!